### ISSUE TITLE
STT0080: finetune stt whisper on grade level 3 AB AND NS.

### ISSUE THREAD

**Issue by gangagyatso4364** [spsither/stt-whisper#4]
### Description 
we have trained the whisper model on 1500 hours of training data for 4.7 epochs after which it showed no progress in validation loss. So now in this task we will continue the finetuning process on grade level 3 of AB AND NS departement training data so that model learns to transcribe audios that are more familiar in everyday use. 

[whisper paper](https://cdn.openai.com/papers/whisper.pdf)
[whisper base model code](https://github.com/openai/whisper)
[whisper  paper vedio](https://www.youtube.com/watch?v=AwJf8aQfChE&t=1825s)

### Completion Criteria 
A trained model on hf that performs better in benchmark specially AB and NS.

### Implementaion
1. set up the ec2 server with necessary specs. 
2. filter grade 3 audios transcription from AB AND NS department.
3. prepare the dataset for training.
4. train the whisper model
5. evaluate the whisper model on benchmark. 
6. create a documentation of model performance comparison.
### Subtask 
- [x] set ssh server with neccessary specs
- [x] filter higher grade 3 audio transcription from AB AND NS 
- [x] run the script to prepare the dataset in arrow format for model training
- [x] train the model on previous checkpoint. 
- [x] save the best performing model checkpoint to hugging face
- [x] evaluate the model 
- [x] create a documentation of model performance.

**Comment by gangagyatso4364** [comment]
```
training time: 1 day. 
cost :20 dollars
training data: 200000
epoch: 1
```

**Comment by gangagyatso4364** [comment]
### **Evaluation Report**

---

1. **Whisper Fine-Tuned on AB and NS**:  
   - **Character Error Rate (CER)**: **25%**  

2. **Whisper Base Model**:  
   - **Character Error Rate (CER)**: **21.1%**

---

#### **Conclusion**:  
- The **Whisper base model** performed **better** than the fine-tuned model on our benchmark test dataset, achieving a **lower CER (21.1%)** compared to the fine-tuned model (**25% CER**).  

- **Possible Reasons**:  
  1. Fine-tuning on AB and NS dialects may have introduced overfitting or reduced generalization to the diverse data in the benchmark test set.

**Comment by gangagyatso4364** [comment]
| Model Name          | Benchmark                                                                                   | Model Type | Benchmark mean CER | STT_AB | STT_CS | STT_MV | STT_NS | STT_TT | STT_HS | STT_NW | STT_PC |
|---------------------|---------------------------------------------------------------------------------------------|------------|--------------------|--------|--------|--------|--------|--------|--------|--------|--------|
| whisper-small-60k   | [benchmark v3](https://huggingface.co/datasets/ganga4364/stt_benchmark)                     | whisper    | 19.75%             | 10.75% | 26.95% | 26.35% | 25.27% | 16.53% | 25.06% | 11.06% | 16.04% |
| whisper-small-95k   | [benchmark v3](https://huggingface.co/datasets/ganga4364/stt_benchmark)                     | whisper    | 19.19%             | 10.80% | 25.92% | 25.91% | 24.54% | 15.37% | 24.38% | 10.78% | 15.81% |
| whisper-small-60k   | [benchmark v1](https://huggingface.co/datasets/openpecha/tibetan-voice-benchmark/resolve/efc6be7d280c458cbad16df58428ae4378706865/test.csv) | whisper    | 21.71%             | 6.05%  | 41.74% | 22.00% | 27.22% | 14.03% |        |        |        |
| whisper-small-95k   | [benchmark v1](https://huggingface.co/datasets/openpecha/tibetan-voice-benchmark/resolve/efc6be7d280c458cbad16df58428ae4378706865/test.csv) | whisper    | 21.24%             | 5.96%  | 40.84% | 21.73% | 27.30% | 14.13% |        |        |        |
| whisper-small-continued   | [benchmark v1](https://huggingface.co/datasets/openpecha/tibetan-voice-benchmark/resolve/efc6be7d280c458cbad16df58428ae4378706865/test.csv) | whisper-finetuned    | 25.15%             | 5.63%   | 44.18%  | 25.75%  | 27.80%  | 15.41%  |        |        |        |
| whisper-small-continued   | [benchmark v3](https://huggingface.co/datasets/ganga4364/stt_benchmark)                     | whisper-finetuned    | 20.52%             | 11.15%  | 27.61%  | 29.68%  | 24.81%  | 16.83%  | 25.69%  | 11.71%  | 16.66%  |

**Comment by gangagyatso4364** [comment]
training data: Grade level 3 from AB and NS department.
|   dept   |  count   |  total_duration_hours  |
|----------|----------|------------------------|
|  STT_AB  |  65864   |         94.12         |
|  STT_NS  | 226060   |        298.58         |

training arguments:
```
training_args = Seq2SeqTrainingArguments(
        output_dir="/data/volume/whisper-small-v5",  # change to a repo name of your choice
        per_device_train_batch_size=8,
        gradient_accumulation_steps=2,  # increase by 2x for every 2x decrease in batch size
        learning_rate=5e-5,
        warmup_steps=500,
        # max_steps=4000,
        num_train_epochs = 1,
        fp16=True,
        evaluation_strategy="steps",
        per_device_eval_batch_size=8,
        predict_with_generate=True,
        generation_max_length=225,
        save_steps=1000,
        eval_steps=1000,
        logging_steps=100,
        save_total_limit=500,
        report_to=["wandb"],
        load_best_model_at_end=True,
        metric_for_best_model="wer",
        greater_is_better=False,
        push_to_hub=True,
        dataloader_num_workers=2,  # Number of workers for DataLoader

    )```

**Comment by gangagyatso4364** [comment]
gcp instance spec:
```ame: ganga
type: ssh-server
image:
  image_uri: >-
    us-docker.pkg.dev/production-01-407505/tfy-docker-us/ssh-server:0.3.7-cu121-py3.11.10
mounts:
  - type: volume
    mount_path: /data/volume
    volume_fqn: tfy-volume://prod-tlib-us-central:tlib-us-ml:ganga-us-volume
resources:
  node:
    type: node_selector
    capacity_type: on_demand
  devices:
    - name: L4
      type: nvidia_gpu
      count: 1
  cpu_limit: 4
  cpu_request: 2
  memory_limit: 160000
  memory_request: 136000
  shared_memory_size: 5000
  ephemeral_storage_limit: 100000
  ephemeral_storage_request: 20000

service_account: bucket-prod-tlib-usctl1-ml-training-artifacts-bucket-sa
home_directory_size: 100
